### PR TITLE
Implemented DIAYN algorithm for skill learning

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
Implemented DIAYN (Diversity is All You Need) algorithm.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -89,6 +90,11 @@ python -m alf.bin.play --root_dir=LOG_DIR
 
   <img src="alf/examples/icm_super_mario_intrinsic_only.png" width = "300" height ="200" alt="super-mario-training-curve"/> <img src="alf/examples/icm_super_mario_intrinsic_only.gif" width = "300" height ="200" alt="super-mario-video"/>
 
+### DIAYN
+* [Pendulum](alf/examples/diayn_pendulum.gin). Learning diverse skills without external reward.
+*
```
Comment (hnyu) [inline README.md]: remove the extra '*'

File: alf/algorithms/agent.py [hunk 2]
```diff
@@ -186,6 +217,12 @@ def train_step(self, exp: Experience, state):
         info = AgentInfo()
         observation = self._encode(exp)
 
+        if self._goal_generator is not None:
```
Comment (hnyu) [inline alf/algorithms/agent.py]: merge two if

File: alf/algorithms/agent.py [hunk 3]
```diff
@@ -186,6 +217,12 @@ def train_step(self, exp: Experience, state):
...
+            new_state = new_state._replace(goal_generator=state.goal_generator)
+
+        if self._goal_generator is not None:
+            observation = [observation, state.goal_generator]
+
```
Comment (hnyu) [inline alf/algorithms/agent.py]: Do you need also to call train_step for the goal generator (if it's trainable)?

File: alf/algorithms/diayn_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            discriminator_net (Network): network for predicting the skill labels based on the observation.
+        """
+        super().__init__(train_state_spec=feature_spec, name=name)
+
+        flat_feature_spec = tf.nest.flatten(feature_spec)
+        assert len(flat_feature_spec
+                   ) == 1, "DIAYNAlgorithm doesn't support nested feature_spec"
+
+        self._num_skills = num_of_skills
+        self._skill_spec = tf.TensorSpec(
```
Comment (hnyu) [inline alf/algorithms/diayn_algorithm.py]: It seems that here you just hard-code the skill_spec to be discrete

File: alf/algorithms/diayn_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        else:
+            skill_discriminate_loss = 0.5 * tf.reduce_mean(
+                tf.square(skill - skill_pred), axis=-1)
+
+        intrinsic_reward = ()
+
+        if calc_intrinsic_reward:
+            # use negative cross-entropy as reward
+            # neglect neg-prior term as it is constant
+            intrinsic_reward = tf.stop_gradient(-skill_discriminate_loss)
```
Comment (hnyu) [inline alf/algorithms/diayn_algorithm.py]: might want to try ScalarAdaptiveNormalizer for normalizing the rewards before returning it

File: alf/algorithms/goal_generator.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import alf.utils.common as common
+
+GoalInfo = namedtuple("GoalInfo", ["goal"], default_value=())
+
+
+@gin.configurable
+class RandomCategoricalGoalGenerator(Algorithm):
+    """Random Goal Generation Module
+
+    This module generates a random categorical goal for the agent.
```
Comment (hnyu) [inline alf/algorithms/goal_generator.py]: "... in the beginning of every episode"

File: alf/algorithms/goal_generator.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    """
+
+    def __init__(self,
+                 num_of_goals,
+                 goal_feature_size,
+                 name="RandomCategoricalGoalGenerator"):
+        """Create a RandomCategoricalGoalGenerator.
+
+        Args:
+            num_of_goals (int): total number of goals the agent can sample from
```
Comment (hnyu) [inline alf/algorithms/goal_generator.py]: comments for other args

File: alf/algorithms/goal_generator.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+@gin.configurable
+class RandomCategoricalGoalGenerator(Algorithm):
+    """Random Goal Generation Module
+
+    This module generates a random categorical goal for the agent.
+    """
+
+    def __init__(self,
+                 num_of_goals,
+                 goal_feature_size,
```
Comment (hnyu) [inline alf/algorithms/goal_generator.py]: I suggest that here we only have one arg goal_spec, and it can be tf.TensorSpec((), tf.int32) or tf.TensorSpec([size], tf.float32) or other form. You can set this spec in a gin file

File: alf/algorithms/agent.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/algorithms/agent.py]: merged

File: README.md (hunk 1 above)
Comment (Haichao-Zhang) [reply README.md]: removed

File: alf/algorithms/agent.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/algorithms/agent.py]: The current goal generator (RandomCategoricalGoalGenerator) is a non-trainable module.
Trainable goal generator will be done in a future PR.

File: alf/algorithms/goal_generator.py (hunk 7 above)
Comment (Haichao-Zhang) [reply alf/algorithms/goal_generator.py]: comments added

File: alf/algorithms/goal_generator.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        samples = tf.reshape(samples, [batch_size, self._goal_feature_size])
+        samples = tf.cast(samples, tf.float32)
+        return samples
+
+    def update_goal(self, observation, state, step_type):
+        new_goal_mask = (step_type == StepType.FIRST)
+        new_goal = common.conditional_update(
+            target=state,
+            cond=new_goal_mask,
+            func=functools.partial(self._generate_goal),
```
Comment (emailweixu) [inline alf/algorithms/goal_generator.py]: no need to use functools.partial here.

File: alf/algorithms/goal_generator.py (hunk 8 above)
Comment (emailweixu) [inline alf/algorithms/goal_generator.py]: If this is always 1 (line 50), why not remove it?

File: alf/algorithms/agent.py [hunk 10]
```diff
@@ -142,6 +154,14 @@ def predict(self, time_step: ActionTimeStep, state: AgentState,
...
 
         new_state = AgentState()
+        if self._goal_generator is not None:
+            goal_step = self._goal_generator.rollout(
+                observation=observation,
+                state=state.goal_generator,
+                step_type=time_step.step_type)
+            new_state = new_state._replace(goal_generator=goal_step.state)
+
+            observation = [observation, goal_step.state]
```
Comment (emailweixu) [inline alf/algorithms/agent.py]: It's better not to use state as goal. It's quite possible that state can contain other information for a different future goal generator. Perhaps goal_step.action (if changing goal generator to be an RLAlgorithm). And we can save the goal to to info so that it can be used at train_step

File: alf/algorithms/diayn_algorithm.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+                 name="DIAYNAlgorithm"):
+        """Create a DIAYNAlgorithm.
+
+        Args:
+            num_of_skills: number of skills
+            hidden_size (int|tuple): size of hidden layer(s)
+            encoding_net (Network): network for encoding observation into a
+                latent feature specified by feature_spec. Its input is same as
+                the input of this algorithm.
+            discriminator_net (Network): network for predicting the skill labels based on the observation.
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: 80c per line. You can set editor.rulers in vscode settings:
{
"editor.rulers": [80]
}

File: alf/algorithms/diayn_algorithm.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        if self._encoding_net is not None:
+            feature, _ = self._encoding_net(observation)
+
+        prev_feature = state
+        skill_index = tf.cast(skill, tf.int32)
+        skill = self._encode_skill(skill_index)
+
+        skill_pred, _ = self._discriminator_net(inputs=feature)
+
+        if tensor_spec.is_discrete(self._skill_spec):
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: self._skill_spec is always discrete at line 61

File: alf/algorithms/diayn_algorithm.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+            outputs=(),
+            state=feature,
+            info=DIAYNInfo(
+                reward=intrinsic_reward,
+                loss=LossInfo(
+                    loss=skill_discriminate_loss,
+                    extra=dict(
+                        skill_discriminate_loss=skill_discriminate_loss))))
+
+    def calc_loss(self, info: DIAYNInfo):
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: do you need to ignore StepType.FIRST?

File: alf/algorithms/goal_generator.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+            observation=observation,
+            state=state)
+        return new_goal
+
+    def rollout(self, observation, state, step_type):
+        """Rollout for one step for goal generation."""
+        new_state = self.update_goal(observation, state, step_type)
+        return PolicyStep(
```
Comment (emailweixu) [inline alf/algorithms/goal_generator.py]: Since you are returning PolicyStep. This class should be subclass of RLAlgorithm instead of Algorithm.

File: alf/algorithms/goal_generator.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+            action=new_state, state=new_state, info=GoalInfo(goal=new_state))
```
Comment (emailweixu) [inline alf/algorithms/goal_generator.py]: Perhaps one_hot encoding for goal will be easier for the policy to learn.

File: alf/algorithms/diayn_algorithm.py [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+                state: observation
+                info (DIAYNInfo):
+        """
+        observations_aug, _ = inputs
+        observation, skill = observations_aug
+
+        if self._encoding_net is not None:
+            feature, _ = self._encoding_net(observation)
+
+        prev_feature = state
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: prev_feature is not used.

File: alf/algorithms/agent.py (hunk 3 above)
Comment (hnyu) [reply alf/algorithms/agent.py]: I think it's still better to leave a train_step entrance here. RandomCategoricalGoalGenerator can just pass this function.

File: alf/algorithms/diayn_algorithm.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+        skill_index = tf.cast(skill, tf.int32)
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: should we use previous skill? Current skill does not affect current observation.

File: alf/algorithms/agent.py [hunk 18]
```diff
@@ -155,9 +175,24 @@ def rollout(self, time_step: ActionTimeStep, state: AgentState, mode):
...
+            new_state = new_state._replace(goal_generator=goal_step.state)
+
+            observation = [observation, goal_step.outputs]
+
         if self._icm is not None:
-            icm_step = self._icm.train_step(
-                (observation, time_step.prev_action), state=state.icm)
+            if self._goal_generator is not None:
+                icm_step = self._icm.train_step(
+                    (observation, time_step.step_type), state=state.icm)
```
Comment (emailweixu) [inline alf/algorithms/agent.py]: Instead of "if-else" here, it's better to make the interface same across the different intrinsic reward algorithms by using time_step as argument.

File: alf/algorithms/agent.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/algorithms/agent.py]: Train_step entrance added.

File: alf/algorithms/diayn_algorithm.py (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/algorithms/diayn_algorithm.py]: Added reward normalizer as an optional input.

File: alf/algorithms/diayn_algorithm.py (hunk 13 above)
Comment (Haichao-Zhang) [reply alf/algorithms/diayn_algorithm.py]: Now the first time step is masked out via valid_masks.

File: alf/algorithms/diayn_algorithm.py [hunk 19]
```diff
@@ -75,54 +78,52 @@ def __init__(self,
...
-                labels=skill, logits=skill_pred)
-        else:
-            skill_discriminate_loss = 0.5 * tf.reduce_mean(
-                tf.square(skill - skill_pred), axis=-1)
+        skill_discriminate_loss = tf.nn.softmax_cross_entropy_with_logits(
+            labels=prev_skill, logits=skill_pred)
+
+        valid_masks = tf.cast(
+            tf.not_equal(step_type, StepType.FIRST), tf.float32)
+        valid_masks = tf.reshape(valid_masks, [batch_size])
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: why do you need reshape here?

File: alf/algorithms/agent.py (hunk 10 above)
Comment (Haichao-Zhang) [reply alf/algorithms/agent.py]: Code has been updated to avoid using goal_step.states as the goal vector directly.

File: alf/algorithms/diayn_algorithm.py (hunk 12 above)
Comment (Haichao-Zhang) [reply alf/algorithms/diayn_algorithm.py]: Unused branch has been removed.

File: alf/algorithms/goal_generator.py (hunk 14 above)
Comment (Haichao-Zhang) [reply alf/algorithms/goal_generator.py]: I have changed the PolicyStep to AlgorithmStep.

File: alf/algorithms/goal_generator.py (hunk 15 above)
Comment (Haichao-Zhang) [reply alf/algorithms/goal_generator.py]: Code has been changed and now the categorical goal generator outputs one_hot encoded goal vectors.

File: alf/algorithms/diayn_algorithm.py [hunk 20]
```diff
@@ -29,37 +30,39 @@
...
+                 reward_normalizer=None,
                  encoding_net: Network = None,
                  discriminator_net: Network = None,
                  name="DIAYNAlgorithm"):
         """Create a DIAYNAlgorithm.
 
         Args:
-            num_of_skills: number of skills
+            num_of_skills (int): number of skills
             hidden_size (int|tuple): size of hidden layer(s)
```
Comment (emailweixu) [inline alf/algorithms/diayn_algorithm.py]: If discriminator_net is None. A default discriminator_net with this hidden_size will be used.

File: alf/algorithms/diayn_algorithm.py (hunk 17 above)
Comment (Haichao-Zhang) [reply alf/algorithms/diayn_algorithm.py]: As the skill is unchanged once generated on the first time step, the "skill" here is essentially the "previous skill", for all time steps except the first one.
Code has been changed to predicting the "previous skills" passed in through state to reduce confusion.

File: alf/algorithms/goal_generator.py [hunk 21]
```diff
@@ -58,22 +56,42 @@ def _generate_goal(self, observation, state):
...
         samples = tf.random.categorical(
             tf.math.log([self._p_goal]), batch_size)
-        samples = tf.reshape(samples, [batch_size, self._goal_feature_size])
-        samples = tf.cast(samples, tf.float32)
-        return samples
+        samples_onehot = tf.one_hot(indices=samples, depth=self._num_of_goals)
+        samples_onehot = tf.reshape(samples_onehot, [batch_size, -1])
+        return samples_onehot
 
     def update_goal(self, observation, state, step_type):
```
Comment (emailweixu) [inline alf/algorithms/goal_generator.py]: prefix private function with `_`

File: alf/algorithms/goal_generator.py [hunk 22]
```diff
@@ -58,22 +56,42 @@ def _generate_goal(self, observation, state):
...
+        new_goal = self.update_goal(observation, state, step_type)
+        return AlgorithmStep(
+            outputs=new_goal,
+            state=GoalState(goal=new_goal),
+            info=GoalInfo(goal=new_goal))
+
+    def train_step(self, inputs, state):
+        """Perform one step of predicting and training computation.
+
+        Note that as RandomCategoricalGoalGenerator is a non-trainable module, this function just passes the goal from state as outputs and
```
Comment (emailweixu) [inline alf/algorithms/goal_generator.py]: line too long

File: alf/algorithms/diayn_algorithm.py (hunk 19 above)
Comment (Haichao-Zhang) [reply alf/algorithms/diayn_algorithm.py]: valid_masks is of the shape [N, 1] while the loss is of shage [N]

File: alf/algorithms/diayn_algorithm.py (hunk 20 above)
Comment (Haichao-Zhang) [reply alf/algorithms/diayn_algorithm.py]: Comments added.

File: alf/algorithms/goal_generator.py (hunk 21 above)
Comment (Haichao-Zhang) [reply alf/algorithms/goal_generator.py]: prefix added.

File: alf/algorithms/goal_generator.py (hunk 22 above)
Comment (Haichao-Zhang) [reply alf/algorithms/goal_generator.py]: format updated

File: alf/algorithms/agent.py [hunk 23]
```diff
@@ -315,8 +310,5 @@ def after_train(self, training_info):
 
     def preprocess_experience(self, exp: Experience):
         reward = self.calc_training_reward(exp.reward, exp.rollout_info)
-        exp_rl = self._rl_algorithm.preprocess_experience(
+        return self._rl_algorithm.preprocess_experience(
             exp._replace(reward=reward, rollout_info=exp.rollout_info.rl))
-        new_rollout_info = exp.rollout_info._replace(rl=exp_rl.rollout_info)
-        exp = exp_rl._replace(rollout_info=new_rollout_info)
-        return exp
```
[the remark this answers is not in the document]
Comment (Haichao-Zhang) [reply alf/algorithms/agent.py]: Now pass the goal vector in a different way therefore this part is not used anymore. Leave this for another PR.

File: alf/algorithms/agent.py [hunk 24]
```diff
@@ -155,9 +175,20 @@ def rollout(self, time_step: ActionTimeStep, state: AgentState, mode):
         new_state = AgentState()
         info = AgentInfo()
         observation = self._encode(time_step)
+
+        if self._goal_generator is not None:
+            goal_step = self._goal_generator.predict(
```
Comment (hnyu) [inline alf/algorithms/agent.py]: Cannot generate costs for on-policy algorithms if the goal generator is trainable

Comment (Haichao-Zhang) [reply alf/algorithms/agent.py]: Updated the interface to handle future on-policy learning as well as trainable goal_generator.
